File: typescript/src/e2b/handle.ts
import type { RunStatusResponse } from "../index.js";
import type { Sandbox } from "./sandbox.js";
import { CommandExitException, type CommandResult } from "./types.js";

const POLL_BASE_MS = 200;
const POLL_MAX_MS = 1_000;

const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

function nextDelay(last: number): number {
  return Math.min(POLL_MAX_MS, Math.max(POLL_BASE_MS, last * 1.5));
}

function decode(bytes: Uint8Array): string {
  return new TextDecoder("utf-8", { fatal: false }).decode(bytes);
}

/** Decode the OpenAPI run-status `stdout`/`stderr` fields (string + encoding) into bytes. */
function decodeStreamBytes(value: unknown, encoding: unknown): Uint8Array {
  if (value instanceof Uint8Array) return value;
  if (typeof value !== "string") return new Uint8Array();
  if (encoding === "base64") {
    if (typeof Buffer !== "undefined") return new Uint8Array(Buffer.from(value, "base64"));
    const binary = atob(value);
    const out = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
    return out;
  }
  return new TextEncoder().encode(value);
}

function readStdout(status: RunStatusResponse): Uint8Array {
  const s = status as { stdout: unknown; stdout_encoding?: unknown };
  return decodeStreamBytes(s.stdout, s.stdout_encoding);
}

function readStderr(status: RunStatusResponse): Uint8Array {
  const s = status as { stderr: unknown; stderr_encoding?: unknown };
  return decodeStreamBytes(s.stderr, s.stderr_encoding);
}

function statusToResult(status: RunStatusResponse): CommandResult {
  return {
    stdout: decode(readStdout(status)),
    stderr: decode(readStderr(status)),
    exitCode: (status as { exit_code?: number }).exit_code ?? -1,
  };
}

export interface WaitOptions {
  onStdout?: (chunk: string) => void;
  onStderr?: (chunk: string) => void;
}

/**
 * Handle for a background run. Live-streaming is approximated by polling
 * `runStatus` and yielding deltas — true streaming requires WS support.
 *
 * TODO(arker-e2b): per-line streaming via the `ws_url` Arker returns from
 * `run`. Needs a WS client in core SDK. See pending-work item #3 in
 * index.ts.
 */
export class CommandHandle {
  readonly pid: number;
  private readonly sbx: Sandbox;
  private readonly runId: string;
  private readonly cmd: string;
  private stdoutEmitted = 0;
  private stderrEmitted = 0;

  constructor(sbx: Sandbox, pid: number, runId: string, cmd: string) {
    this.sbx = sbx;
    this.pid = pid;
    this.runId = runId;
    this.cmd = cmd;
  }

  async wait(opts: WaitOptions = {}): Promise<CommandResult> {
    let delay = POLL_BASE_MS;
    let last: RunStatusResponse | undefined;
    while (true) {
      const status = await this.sbx._computer.runStatus(this.runId);
      this.fireDeltas(status, opts);
      if ((status as { completed?: boolean }).completed) {
        last = status;
        break;
      }
      await sleep(delay);
      delay = nextDelay(delay);
    }
    const result = statusToResult(last!);
    if (result.exitCode !== 0) throw new CommandExitException(result);
    return result;
  }

  async kill(): Promise<boolean> {
    try {
      const r = await this.sbx._computer.cancelRun(this.runId);
      return !!r.cancelled;
    } finally {
      this.sbx._forgetPid(this.pid);
    }
  }

  disconnect(): void {
    this.sbx._forgetPid(this.pid);
  }

  async *iter(): AsyncIterableIterator<string> {
    let delay = POLL_BASE_MS;
    while (true) {
      const status = await this.sbx._computer.runStatus(this.runId);
      const stdout = readStdout(status);
      const newOut = stdout.subarray(this.stdoutEmitted);
      if (newOut.length > 0) {
        this.stdoutEmitted = stdout.length;
        yield decode(newOut);
      }
      if ((status as { completed?: boolean }).completed) return;
      await sleep(delay);
      delay = nextDelay(delay);
    }
  }

  [Symbol.asyncIterator](): AsyncIterableIterator<string> {
    return this.iter();
  }

  private fireDeltas(status: RunStatusResponse, opts: WaitOptions): void {
    const stdout = readStdout(status);
    const stderr = readStderr(status);
    if (opts.onStdout) {
      const d = stdout.subarray(this.stdoutEmitted);
      if (d.length > 0) opts.onStdout(decode(d));
    }
    if (opts.onStderr) {
      const d = stderr.subarray(this.stderrEmitted);
      if (d.length > 0) opts.onStderr(decode(d));
    }
    this.stdoutEmitted = stdout.length;
    this.stderrEmitted = stderr.length;
  }
}

File: python/src/arker/e2b/_handle.py
"""`CommandHandle` — return type of `commands.run(..., background=True)`.

Wraps an Arker background run_id behind the e2b CommandHandle API:
`.pid`, `.wait()`, `.kill()`, `.disconnect()`, `__iter__`.

TODO(arker-e2b): true per-line `on_stdout` / `on_stderr` streaming.
Today `wait()` and `__iter__` poll `run_status` and emit whatever delta
arrived since the last poll — coarser than e2b's per-line WS push. Real
streaming uses the `ws_url` Arker already returns; needs a WS client
in core SDK. See pending-work item #3 in __init__.py.
"""

from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING, Callable, Iterator

from ..computer import RunStatusResponse
from ._types import CommandExitException, CommandResult

if TYPE_CHECKING:
    from ._sandbox import Sandbox

logger = logging.getLogger("arker.e2b")

POLL_BASE_S = 0.2
POLL_MAX_S = 1.0


def _status_to_command_result(status: RunStatusResponse) -> CommandResult:
    return CommandResult(
        stdout=status.stdout.decode("utf-8", errors="replace"),
        stderr=status.stderr.decode("utf-8", errors="replace"),
        exit_code=status.exit_code if status.exit_code is not None else -1,
    )


def _next_delay(last: float) -> float:
    return min(POLL_MAX_S, max(POLL_BASE_S, last * 1.5))


class CommandHandle:
    def __init__(self, sandbox: "Sandbox", pid: int, run_id: str, cmd: str) -> None:
        self._sandbox = sandbox
        self._pid = pid
        self._run_id = run_id
        self._cmd = cmd
        self._stdout_emitted = 0
        self._stderr_emitted = 0

    @property
    def pid(self) -> int:
        return self._pid

    def wait(
        self,
        on_stdout: Callable[[str], None] | None = None,
        on_stderr: Callable[[str], None] | None = None,
        on_pty: Callable | None = None,
    ) -> CommandResult:
        """Block until the run finishes; fire callbacks for delta chunks."""
        delay = POLL_BASE_S
        last_status: RunStatusResponse | None = None
        while True:
            status = self._sandbox._computer.run_status(self._run_id)
            self._fire_deltas(status, on_stdout, on_stderr)
            if status.completed:
                last_status = status
                break
            time.sleep(delay)
            delay = _next_delay(delay)

        result = _status_to_command_result(last_status)
        if result.exit_code != 0:
            raise CommandExitException(result)
        return result

    def kill(self) -> bool:
        try:
            return bool(self._sandbox._computer.cancel_run(self._run_id).cancelled)
        finally:
            self._sandbox._forget_pid(self._pid)

    def disconnect(self) -> None:
        self._sandbox._forget_pid(self._pid)

    def __iter__(self) -> Iterator[str]:
        """Yield stdout deltas until the run completes."""
        delay = POLL_BASE_S
        while True:
            status = self._sandbox._computer.run_status(self._run_id)
            new_stdout = status.stdout[self._stdout_emitted :]
            if new_stdout:
                self._stdout_emitted = len(status.stdout)
                yield new_stdout.decode("utf-8", errors="replace")
            if status.completed:
                return
            time.sleep(delay)
            delay = _next_delay(delay)

    def _fire_deltas(
        self,
        status: RunStatusResponse,
        on_stdout: Callable[[str], None] | None,
        on_stderr: Callable[[str], None] | None,
    ) -> None:
        if on_stdout:
            new_out = status.stdout[self._stdout_emitted :]
            if new_out:
                on_stdout(new_out.decode("utf-8", errors="replace"))
        if on_stderr:
            new_err = status.stderr[self._stderr_emitted :]
            if new_err:
                on_stderr(new_err.decode("utf-8", errors="replace"))
        self._stdout_emitted = len(status.stdout)
        self._stderr_emitted = len(status.stderr)
